cdn: add SignCookie example w/ tests

elithrar opened the pull request:
This PR adds support for the upcoming Signed Cookies & URL prefixes feature, which is currently in closed alpha. These extend our current [Signed URL](https://cloud.google.com/cdn/docs/using-signed-urls) functionality.

- [x] Example added for `SignCookie`
- [x] Tests added for `SignCookie`
- [x] Updated the existing `SignURL`  docstring to be uniform across the two examples
- [x] The SignURL test cases now use subtests, resolving a TODO from ahmetb@
- [x] Example added for `SignURLWithPrefix`
- [x] Tests added for `SignURLWithPrefix`

elithrar commented:
* Added `SignURLWithPrefix` and relevant tests
* Cleaned up tests for existing `SignURL` for uniformity.

PTAL.

elithrar commented:
@ahmetb - not sure if still in scope for you, but as you wrote the original impl, if you can cast an eye over this, it'd be great.

I've kept the examples (URL vs. Cookie) self-contained at the expense of some duplication, which felt worthwhile. We'll separate them in the docs so that users can understand the implementation for each easily.

ahmetb commented:
Signed URLs stuff was a one-off for me and I don’t really work on this stuff anymore. But took a peek and looks reasonable.

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+		Name:   "Cloud-CDN-Cookie",
+		Value:  signedValue,
+		Path:   "/segments/", // Best practice: only send the cookie for paths it is valid for
+		Domain: "media.example.com",
```
Use the vars from above?

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+// - key should be in raw form (not base64url-encoded) which is
+// 16-bytes long.
+// - keyName must match a key added to the backend service or bucket.
+func SignCookie(urlPrefix, keyName string, key []byte, expiration time.Time) (string, error) {
```
A little different when this is a quickstart versus a standalone sample. But, we generally try to minimize args for standalone samples. We also give sample values inside the function body to make it easy to copy out.

https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#function-arguments-for-snippets

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+	// We print this to stdout in this example.
+	// Use http.ResponseWriter.SetCookie to write a cookie to an authenticated
+	// client in a real application.
+	fmt.Println(cookie)
```
If you switch this to a standalone sample, print to an `io.Writer`. 

https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#print-to-an-iowriter-for-snippets

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,140 @@
+		t.Fatal(err)
+	}
+	if !bytes.Equal(b, expected) {
+		t.Fatalf("got=%v; expected=%v", b, expected)
```
Include the name of the thing that failed in the failure message. Same for the other failures.

https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,140 @@
+		keyName    string
+		key        []byte
+		expiration time.Time
+		out        string
```
This is usually named `want` or `expected`.

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,140 @@
+		key        []byte
+		expiration time.Time
+		out        string
+		shouldPass bool
```
Maybe `shouldMatch`? Wasn't clear if this was for errors or something else at first.

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,140 @@
+		shouldPass bool
+	}{
+		{
+			// Valid signature
```
Probably don't need these comments -- the `testName` should be enough for docs.

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	for _, c := range cases {
+		t.Run(
```
One line. Could declare the test name separately if it is too long. The indentation is a little hard for me to follow with this extra line break.

tbpg commented on cdn/signedurls/signurl_test.go:
```diff
@@ -45,27 +45,90 @@ func TestReadKeyFile(t *testing.T) {
-		if signed != c.out {
-			t.Errorf("want=%v\ngot=%v", c.out, signed)
-		}
+		t.Run("%s", func(t *testing.T) {
```
```suggestion
		t.Run(c/testName, func(t *testing.T) {
```

elithrar commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+	"time"
+)
+
+// [START example]
```
Noted - I'd copied this from the existing example. Will prefix as `cdn_signedcookie_example` and similar for others.

elithrar commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+// - key should be in raw form (not base64url-encoded) which is
+// 16-bytes long.
+// - keyName must match a key added to the backend service or bucket.
+func SignCookie(urlPrefix, keyName string, key []byte, expiration time.Time) (string, error) {
```
(Same as before - followed the previous examples. I'm all for keeping these as private to avoid dependendencies)

elithrar commented:
Updates:

* Updated the region tag for the _new_ Signed Cookie example only. Left the existing tags as-is as to not break the docs.
* Moved imports into the region tags
* `package main` -> `package <featurename>`
* Renamed `main()` to `example()` for simplicity to retain the function usage examples
* Unexported all functions
* Some formatting fixes to address comments

@tbpg - please take a look when ready - thanks!

elithrar commented:
(Fixed some of the test outputs that had a typo )

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+
+func main() {
+	var keyPath string
+	flag.StringVar(&keyPath, "key-file", "", "The path to a file containing the base64-encoded signing key")
```
Looks like this is still open.

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+	return d[:n], nil
+}
+
+func example(w io.Writer) {
```
More descriptive name than `example`?

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	// We print this to stdout in this example.
+	// Use http.ResponseWriter.SetCookie to write a cookie to an authenticated
```
Could we make this function an `http.HandlerFunc`?

tbpg commented on cdn/signedcookies/sign_cookie_test.go:
```diff
@@ -0,0 +1,131 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package signedcookie
```
Newline above.

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+		expiration = time.Hour * 2
+	)
+
+	signedValue, err := signCookie(
```
Check error.

tbpg commented on cdn/signedcookies/sign_cookie.go:
```diff
@@ -0,0 +1,120 @@
+
+	key, err := readKeyFile(keyPath)
+	if err != nil {
+		log.Fatal(err)
```
Return errors (or use `http.Error` if you convert this to a handler).

https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#return-errors

elithrar commented:
Sent an email to discuss how we tackle adding to existing code samples.

elithrar commented:
@tbpg - comments addressed. LMK if anything else is needed.

elithrar commented:
Thanks Tyler!

>
